src/statsmodels/formula.jl

```diff
@@ -291,57 +412,105 @@ function nc(trm::Vector)
+    mm = hcat([expandcols(tc) for tc in mm_cols]...)
+    mm_col_term_nums = vcat([fill(i_term, nc(tc)) for (i_term,tc) in enumerate(mm_cols)]...)
+
+    ModelMatrix{Float64}(mm, mm_col_term_nums)
+
 end
 
+
+
 termnames(term::Symbol, col) = [string(term)]
-function termnames(term::Symbol, col::PooledDataArray)
-    levs = levels(col)
-    [string(term, " - ", levs[i]) for i in 2:length(levs)]
+function termnames(term::Symbol, mf::ModelFrame; non_redundant::Bool = false)
+    if haskey(mf.contrasts, term)
+        termnames(term, mf.df[term], 
+             non_redundant ? promote_contrast(mf.contrasts[term]) : mf.contrasts[term])
+    else
+        termnames(term, mf.df[term])
+    end
 end
+termnames(term::Symbol, col::Any, contrast::ContrastsMatrix) =
+    ["$term: $name" for name in contrast.termnames]
 
-function coefnames(mf::ModelFrame)
-    if mf.terms.intercept
-        vnames = UTF8String["(Intercept)"]
+
+function expandtermnames(term::Vector)
+    if length(term) == 1
+        return term[1]
     else
-        vnames = UTF8String[]
-    end
-    # Need to only include active levels
-    for term in mf.terms.terms
-        if isa(term, Expr)
-            if term.head == :call && term.args[1] == :|
-                continue                # skip random-effects terms
-            elseif term.head == :call && term.args[1] == :&
-                ## for an interaction term, combine term names pairwise,
-                ## starting with rightmost terms
-                append!(vnames,
-                        foldr((a,b) ->
-                              vec([string(lev1, " & ", lev2) for
+        return foldr((a,b) -> vec([string(lev1, " & ", lev2) for
                                    lev1 in a,
                                    lev2 in b]),
-                              map(x -> termnames(x, mf.df[x]), term.args[2:end])))
-            else
-                error("unrecognized term $term")
-            end
-        else
-            append!(vnames, termnames(term, mf.df[term]))
+                     term)
+    end
+end
+
+function coefnames(mf::ModelFrame)
+    ## strategy mirrors ModelMatrx constructor:
+    eterm_names = Dict{Tuple{Symbol,Bool}, Vector{String}}()
+    term_names = Vector{Vector{String}}[]
+    mf.terms.intercept && push!(term_names, Vector[String["(Intercept)"]])
+
+    factors = round(Bool, mf.terms.factors)
+
+    for (i_term, term) in enumerate(mf.terms.terms)
+        isfe(term) || continue
+        ## names for columns for eval terms
+        names = Vector{String}[]
+
+        i_term += mf.terms.response
+        eterms = mf.terms.eterms[factors[:, i_term]]
+        non_redundant = mf.non_redundant_terms[factors[:, i_term], i_term]
+
+        for et_and_nr in zip(eterms, non_redundant)
+            haskey(eterm_names, et_and_nr) ||
+            setindex!(eterm_names,
+                      termnames(et_and_nr[1], mf, non_redundant=et_and_nr[2]),
```

Comment:
This only calls `termnames` once for each eval term (column in the data frame), or twice if it occurs in both redundant and non-redundant forms, regardless of how many terms it appears in.

That being said, it might make sense to calculate the coef names once and store them on the model matrix itself if we switch to something like a `NamedArray`. Then some of this could be refactored into the same loop. But the issue then would be that `coefnames` lives on the model matrix, whereas now it lives on the model frame, so the interface would have to change. That would (as far as I can tell) be pretty minimal, since the only place we use it is in `coeftable` for `DataFrameModels`, which hold the `ModelMatrix` and the `ModelFrame`.